src/core/featuremodel.cpp

```diff
@@ -351,58 +351,147 @@ void FeatureModel::featureAdded( QgsFeatureId fid )
-    mLayer->addTopologicalPoints( mFeature.geometry() );
+  bool isSuccess = true;
+  connect( mLayer, &QgsVectorLayer::featureAdded, this, &FeatureModel::featureAdded );
 
-  if ( commit() )
+  if ( mLayer->addFeature( mFeature ) )
   {
-    QgsFeature feat;
-    if ( mLayer->getFeatures( QgsFeatureRequest().setFilterFid( mFeature.id() ) ).nextFeature( feat ) )
-      setFeature( feat );
+    if ( QgsProject::instance()->topologicalEditing() )
+      mLayer->addTopologicalPoints( mFeature.geometry() );
+
+    if ( commit() )
+    {
+      QgsFeature feat;
+      if ( mLayer->getFeatures( QgsFeatureRequest().setFilterFid( mFeature.id() ) ).nextFeature( feat ) )
+      {
+        setFeature( feat );
+      }
+      else
+      {
+        QgsMessageLog::logMessage( tr( "Feature %1 could not be fetched after commit" ).arg( mFeature.id() ), "QField", Qgis::Warning );
+        isSuccess = false;
+      }
+    }
     else
-      QgsMessageLog::logMessage( tr( "Feature %1 could not be fetched after commit" ).arg( mFeature.id() ), "QField", Qgis::Warning );
+    {
+      QgsMessageLog::logMessage( tr( "Feature %1 could not be committed" ).arg( mFeature.id() ), "QField", Qgis::Warning );
+      isSuccess = false;
+    }
+  }
+  else
+  {
+    QgsMessageLog::logMessage( tr( "Feature could not be added" ), "QField", Qgis::Critical );
+    isSuccess = false;
   }
+
   disconnect( mLayer, &QgsVectorLayer::featureAdded, this, &FeatureModel::featureAdded );
+  return isSuccess;
 }
 
-void FeatureModel::deleteFeature()
+bool FeatureModel::deleteFeature()
 {
-  startEditing();
+  if ( ! startEditing() )
+  {
+    QgsMessageLog::logMessage( tr( "Cannot start editing" ), "QField", Qgis::Critical );
+    return false;
+  }
 
   //delete child features in case of compositions
   const QList<QgsRelation> referencingRelations = QgsProject::instance()->relationManager()->referencedRelations( mLayer );
+  QList<QgsVectorLayer *> childLayersEdited;
+  bool isSuccess = true;
   for ( const QgsRelation &referencingRelation : referencingRelations )
   {
     if ( referencingRelation.strength() == QgsRelation::Composition )
     {
       QgsVectorLayer *childLayer = referencingRelation.referencingLayer();
-      childLayer->startEditing();
-      QgsFeatureIterator relatedFeaturesIt = referencingRelation.getRelatedFeatures( mFeature );
-      QgsFeature childFeature;
-      while ( relatedFeaturesIt.nextFeature( childFeature ) )
+      
+      if ( childLayer->startEditing() )
       {
-        childLayer->deleteFeature( childFeature.id() );
+        QgsFeatureIterator relatedFeaturesIt = referencingRelation.getRelatedFeatures( mFeature );
+        QgsFeature childFeature;
+        while ( relatedFeaturesIt.nextFeature( childFeature ) )
+        {
+          if ( ! childLayer->deleteFeature( childFeature.id() ) )
+          {
+            QgsMessageLog::logMessage( tr( "Cannot delete feature from child layer" ), "QField", Qgis::Critical );
```

Comment:
As well here maybe the child layer name in the message.

Reply:
Done